op/model.py

```diff
@@ -204,26 +215,29 @@ def __getitem__(self, key):
 
 # We mix in MutableMapping here to get some convenience implementations, but whether it's actually
 # mutable or not is controlled by the flag.
-class RelationUnitData(LazyMapping, MutableMapping):
-    def __init__(self, relation, unit, is_mutable, backend):
+class RelationDataContent(LazyMapping, MutableMapping):
+    def __init__(self, relation, entity, is_mutable, backend):
         self.relation = relation
-        self.unit = unit
+        self._entity = entity
         self._is_mutable = is_mutable
         self._backend = backend
+        self._is_app = isinstance(entity, Application)
 
     def _load(self):
         try:
-            return self._backend.relation_get(self.relation.id, self.unit.name)
+            return self._backend.relation_get(self.relation.id, self._entity.name, self._is_app)
         except RelationNotFoundError:
             # Dead relations tell no tales (and have no data).
             return {}
 
     def __setitem__(self, key, value):
-        if not self._is_mutable:
-            raise RelationDataError(f'cannot set relation data for {self.unit.name}')
+        if not self._is_mutable(self._backend):
```

Comment:
Sorry for not noticing this earlier. This is outsourcing local knowledge from this type into the constructor. That's what's creating the awkwardness mentioned earlier. We need to move the logic into this type instead of putting everything into the constructor, including a lambda which is really just logic pertaining to this type.

Reply:
Agreed, there is a better way to do it now that we have `unit_name` and `app_name` attributes available from the `ModelBackend` type. Previously this was only accessible through the model which we do not have a reference to.